conversion/conversion.go

```diff
@@ -201,27 +255,85 @@ func snapshotMigrationHandler(sourceProfile profiles.SourceProfile, config write
 }
 
 func dataFromDatabase(ctx context.Context, sourceProfile profiles.SourceProfile, targetProfile profiles.TargetProfile, config writer.BatchWriterConfig, conv *internal.Conv, client *sp.Client) (*writer.BatchWriter, error) {
-	infoSchema, err := GetInfoSchema(sourceProfile, targetProfile)
-	if err != nil {
-		return nil, err
-	}
-	var streamInfo map[string]interface{}
-	if sourceProfile.Conn.Streaming {
-		streamInfo, err = infoSchema.StartChangeDataCapture(ctx, conv)
-		if err != nil {
-			return nil, err
+	//handle migrating data for sharded migrations differently
+	//sharded migrations are identified via the config= flag, if that flag is not present
+	//carry on with the existing code path in the else block
+	switch sourceProfile.Ty {
+	case profiles.SourceProfileTypeConfig:
+		////There are three cases to cover here, bulk migrations and sharded migrations (and later DMS)
+		//We provide an if-else based handling for each within the sharded code branch
+		//This will be determined via the configType, which can be "bulk", "dataflow" or "dms"
+		if sourceProfile.Config.ConfigType == constants.BULK_MIGRATION {
+			var bw *writer.BatchWriter
+
+			//Migrate the data from the data shards, the schema shard needs to be specified here again.
+			for _, dataShard := range sourceProfile.Config.ShardConfigurationBulk.DataShards {
+				//Create a connection profile object for it
+				fmt.Printf("Initiating migration for shard: %v\n", dataShard.DbName)
+				infoSchema, err := getInfoSchemaForShard(dataShard, sourceProfile.Driver, targetProfile)
+				if err != nil {
+					return nil, err
+				}
+				//finally perform a snapshot migration for the shard
+				bw = performSnapshotMigration(config, conv, client, infoSchema)
+			}
+			//finally, once all shard migrations are complete, return the batch writer object
+			return bw, nil
+		} else if sourceProfile.Config.ConfigType == constants.DATAFLOW_MIGRATION {
+			//STEP - 1 - Create batch for each physical shard
+			asyncProcessShards := func(p *profiles.DataShard, mutex *sync.Mutex) common.TaskResult[*profiles.DataShard] {
+				//create streaming cfg from the config source type.
+				streamingCfg := streaming.CreateStreamingConfig(*p)
```

Comment:
How does this handle if the schema is different on various Datashards ? i.e. if there are tables which need to be configured as per schema source, but don't exist on the data shard ?

Reply:
Since we do not connect to the data shards from Harbourbridge in the case of dataflow based migrations, there is no way to know if the tables which have been configured from the schema source are present in the data source. 
This would need to be a guarantee coming from the customer. 

That being said, let's list out the scenarios - 

1. Identical schema - Happy path, proceed ahead.
2. Table X exists in Data shard but does not exist in schema source - This table will not be part of the session file. Correspondingly, no table will be created for it in Spanner. The CDC events which will come for this table will be  ignored.
3. Table X does not exist in data shard but exist in schema source - It will be created in Spanner. For data shards, which have this table, it will be streamed and populated. Data shards which do not have this table will not generate CDC for it and not contribute to writing to Spanner for the Table X.

Reply:
Got it, that error has more to do with individual table migrations than sharding. Currently, the `dbName` coming from the sharding configuration is specified in the include list during datastream configuration.

Will have to check out individual table migrations + sharding work together.